crates/bevy_render/src/mesh/allocator.rs

```diff
@@ -0,0 +1,1025 @@
+/// [`OffsetAllocator`]. Slabs start small and then grow as their contents fill
+/// up, up to a maximum size limit. To reduce fragmentation, vertex and index
+/// buffers that are too large bypass this system and receive their own buffers.
+///
+/// The [`MeshAllocatorSettings`] allows you to tune the behavior of the
+/// allocator for better performance with your application. Most applications
+/// won't need to change the settings from their default values.
+#[derive(Resource)]
+pub struct MeshAllocator {
+    /// Holds all buffers and allocators.
+    slabs: HashMap<SlabId, Slab>,
+
+    /// Maps a layout to the slabs that hold elements of that layout.
+    ///
+    /// This is used when allocating, so that we can find the appropriate slab
+    /// to place an object in.
+    slab_layouts: HashMap<ElementLayout, Vec<SlabId>>,
+
+    /// Maps mesh asset IDs to the ID of the slabs that hold their vertex data.
+    mesh_id_to_vertex_slab: HashMap<AssetId<Mesh>, SlabId>,
+
+    /// Maps mesh asset IDs to the ID of the slabs that hold their index data.
+    mesh_id_to_index_slab: HashMap<AssetId<Mesh>, SlabId>,
+
+    /// The next slab ID to assign.
+    next_slab_id: SlabId,
+
+    /// Whether we can pack multiple vertex arrays into a single slab on this
+    /// platform.
+    ///
+    /// This corresponds to [`DownlevelFlags::BASE_VERTEX`], which is unset on
+    /// WebGL 2. On this platform, we must give each vertex array its own
+    /// buffer, because we can't adjust the first vertex when we perform a draw.
+    general_vertex_slabs_supported: bool,
+}
+
+/// Tunable parameters that customize the behavior of the allocator.
+///
+/// Generally, these parameters adjust the tradeoff between memory fragmentation
+/// and performance. You can adjust them as desired for your application. Most
+/// applications can stick with the default values.
+#[derive(Resource)]
+pub struct MeshAllocatorSettings {
+    /// The minimum size of a slab (hardware buffer), in bytes.
+    ///
+    /// The default value is 1 MB.
+    pub min_slab_size: u64,
+
+    /// The maximum size of a slab (hardware buffer), in bytes.
+    ///
+    /// When a slab reaches this limit, a new slab is created.
+    ///
+    /// The default value is 512 MB.
+    pub max_slab_size: u64,
+
+    /// The maximum size of vertex or index data that can be placed in a general
+    /// slab, in bytes.
+    ///
+    /// If a mesh has vertex or index data that exceeds this size limit, that
+    /// data is placed in its own slab. This reduces fragmentation, but incurs
+    /// more CPU-side binding overhead when drawing the mesh.
+    ///
+    /// The default value is 256 MB.
+    pub large_threshold: u64,
+
+    /// The factor by which we scale a slab when growing it.
+    ///
+    /// This value must be greater than 1. Higher values result in more
+    /// fragmentation but fewer expensive copy operations when growing the
+    /// buffer.
+    ///
+    /// The default value is 1.5.
```

Comment:
Why not 2?

Reply:
See https://archive.ph/Z2R8w. tl;dr a value closer to the golden ratio increases the chance that if you reallocate more than once you're able to coalesce older spaces into the new space.